spec/main_spec.rb

```diff
@@ -0,0 +1,233 @@
+require './main.rb'
+
+RSpec.describe Enumerable do
+  let(:arr) { [1, 2, 3, 4, 5] }
+  let(:arr_regex) { %w[cat bat rat hello] }
+  let(:arr_nil) { [1, nil, false] }
+  let(:match_arr) { [1, 1, 1] }
+  let(:arr_empty) { [] }
+  let(:arr_count) { [1, 1, 1, 2, 3] }
+
+  describe '#my_each' do
+    it 'should return Enumerator if block is not given' do
+      expect(arr.my_each).to be_a(Enumerator)
+      expect(arr.my_each.to_a).to eq(arr)
+    end
+
+    it 'should not use the built-in Array#each' do
+      expect(arr).to_not receive(:each)
+    end
+
+    it 'should work as #each method if block given' do
+      expect(arr.my_each { |ele| puts "return: #{ele}" }).to eq(arr.each { |ele| puts "return: #{ele}" })
+    end
+
+    it 'should take 0 arguments' do
+      expect { arr.my_each('argument') }.to raise_error(ArgumentError)
+      expect { arr.my_each }.not_to raise_error(ArgumentError)
```

Comment:
`OPTIONAL` WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`.